pkg/scheduler/plugins/gpushare/utils.go

```diff
@@ -0,0 +1,224 @@
+// nvidia.com/gpu means applying for full-card
+// koordinator.sh/gpu means applying for cards in percentile
+// convertGPUResource will convert either nvidia.com/gpu or koordinator.sh/gpu to koordinator.sh/gpu-core and koordinator.sh/gpu-memory-ratio
+func convertGPUResource(podRequest corev1.ResourceList, combination uint) corev1.ResourceList {
+	switch combination {
+	case gpuCoreExist | gpuMemoryExist:
+		return corev1.ResourceList{
+			apiext.GPUCore:   podRequest[apiext.GPUCore],
+			apiext.GPUMemory: podRequest[apiext.GPUMemory],
+		}
+	case gpuCoreExist | gpuMemoryRatioExist:
+		return corev1.ResourceList{
+			apiext.GPUCore:        podRequest[apiext.GPUCore],
+			apiext.GPUMemoryRatio: podRequest[apiext.GPUMemoryRatio],
+		}
+	case koordGpuExist:
+		return corev1.ResourceList{
+			apiext.GPUCore:        podRequest[apiext.KoordGPU],
+			apiext.GPUMemoryRatio: podRequest[apiext.KoordGPU],
+		}
+	case nvidiaGpuExist:
+		nvidiaGpu := podRequest[apiext.NvidiaGPU]
+		return corev1.ResourceList{
+			apiext.GPUCore:        *resource.NewQuantity(nvidiaGpu.Value()*100, resource.DecimalSI),
+			apiext.GPUMemoryRatio: *resource.NewQuantity(nvidiaGpu.Value()*100, resource.DecimalSI),
+		}
+	}
+	return nil
+}
+
+func checkNodeGPUSatisfied(podRequest corev1.ResourceList, nodeDeviceFree *deviceResources) bool {
+	if isMultipleGPUPod(podRequest) {
+		gpuCore, gpuMem, gpuMemRatio := podRequest[apiext.GPUCore], podRequest[apiext.GPUMemory], podRequest[apiext.GPUMemoryRatio]
+		gpuWanted := gpuCore.Value() % 100
+		podRequestPerCard := corev1.ResourceList{
+			apiext.GPUCore:        *resource.NewQuantity(gpuCore.Value()/gpuWanted, resource.DecimalSI),
+			apiext.GPUMemory:      *resource.NewQuantity(gpuMem.Value()/gpuWanted, resource.BinarySI),
+			apiext.GPUMemoryRatio: *resource.NewQuantity(gpuMemRatio.Value()/gpuWanted, resource.DecimalSI),
+		}
+
+		satisfiedDeviceCount := 0
+		for _, resources := range *nodeDeviceFree {
+			if satisfiedDeviceCount == int(gpuWanted) {
+				break
+			}
+			if satisfied, _ := quotav1.LessThanOrEqual(podRequestPerCard, resources); satisfied {
+				satisfiedDeviceCount++
+			}
+		}
+		return satisfiedDeviceCount == int(gpuWanted)
+	}
+
+	for _, resources := range *nodeDeviceFree {
+		if satisfied, _ := quotav1.LessThanOrEqual(podRequest, resources); satisfied {
+			return true
+		}
+	}
+
+	return false
+}
+
+func isMultipleGPUPod(podRequest corev1.ResourceList) bool {
+	gpuCore := podRequest[apiext.GPUCore]
+	return gpuCore.Value() > 100 && gpuCore.Value()%100 == 0
+}
+
+func memRatioToBytes(ratio, totalMemory resource.Quantity) resource.Quantity {
+	return *resource.NewQuantity(ratio.Value()*totalMemory.Value()/100, resource.BinarySI)
+}
+
+func memBytesToRatio(bytes, totalMemory resource.Quantity) resource.Quantity {
+	return *resource.NewQuantity(bytes.Value()/totalMemory.Value()*100, resource.DecimalSI)
+}
+
+func allocateGPU(podRequest corev1.ResourceList, nodeDeviceFree *deviceResources) []*apiext.DeviceAllocation {
```

Comment:
allocateGPU and checkNodeGPUSatisfied is very similar. checkNodeGPUSatisfied can rename to tryAssign, if not satisfy, return nil, else return allocate result like allocateGPU.